src/Preact10ShallowDiff.ts

```diff
@@ -0,0 +1,759 @@
+      unmount(this._prevElement);
+    }
+
+    if (this._memoElement) {
+      unmount(this._memoElement);
+    }
+
+    Preact10ShallowDiff.current = null;
+    this._reset();
+  }
+
+  private _reset() {
+    this._prevElement = null;
+    this._memoElement = null;
+    this._componentInstance = null;
+    this._rendered = null;
+
+    Preact10ShallowDiff.current = null;
+  }
+
+  private _diffComponent(
+    newVNode: ComponentVNode<any>,
+    oldVNode: ComponentVNode<any>,
+    globalContext: any,
+    commitQueue: any[]
+  ) {
+    // React's memo function wraps a component. It doesn't create a new VNode in
+    // the virtual tree. To be compatible with React's shallow renderer, we'll
+    // try to mimic this behavior so shallow enzyme tests don't need to be aware
+    // of this subtle difference between Preact and React. Further, how Preact
+    // implements Memo today may change in the future so we'll try to avoid
+    // exposing that implementation detail to users.
+    //
+    // To mimic React's behavior here, we'll render through memo components. If
+    // `render` is given an element representing a Memoed component, we will
+    // render Memo and the component that it memos (assuming the Memo doesn't
+    // instruct us to skip updating). The public methods on this class should
+    // never return references to the Memo component, but always the component
+    // that it is memoizing.
+
+    let renderResult: ComponentChild | typeof skipUpdateSymbol;
+    if (isMemo(newVNode)) {
+      this._memoElement = newVNode;
+      renderResult = diffComponent(
+        newVNode,
+        oldVNode,
+        globalContext,
+        commitQueue
+      );
+
+      if (renderResult == skipUpdateSymbol) {
+        // The memo component told us not to update, so return the result of the
+        // previous render
+        return this._rendered;
+      }
+
+      if (isComponentVNode(renderResult)) {
+        // The memo component updated, so let's setup to update and diff the
+        // component it returned
+        newVNode = renderResult;
+        oldVNode = this._prevElement ?? ({} as any);
+      } else {
+        throw new Error(
+          `Memo rendered a non-component element. Only memo'ed components is currently supported for shallow rendering`
+        );
+      }
+    }
+
+    this._prevElement = newVNode;
+    renderResult = diffComponent(
+      newVNode,
+      oldVNode,
+      globalContext,
+      commitQueue
+    );
+
+    // If diffing this component told us not to update, then return the result
+    // of the previous render. Else, return the result of the diff
+    return renderResult === skipUpdateSymbol ? this._rendered : renderResult;
+  }
+}
+
+function isComponentVNode<P = any>(element: any): element is ComponentVNode<P> {
+  return isValidElement(element) && typeof element.type === 'function';
+}
+
+function assertIsComponentVNode(
+  element: JSX.Element
+): asserts element is ComponentVNode {
```

Comment:
TIL about [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions) in TypeScript.

Reply:
Yea! They are pretty neat :)